unixfs/hamt/fetcher.go

```diff
@@ -0,0 +1,297 @@
+	doneCnt    int // job's done but results not yet retrieved
+	hits       int // job's result already ready, no delay
+	nearMisses int // job currently being worked on, small delay
+	misses     int // job on todo stack but will be done in the next batch, larger delay
+
+	// other useful stats
+	cidCnt int
+
+	start time.Time
+}
+
+// batchSize must be at least as large as the largest number of cids
+// requested in a single job.  For best perforamce it should likely be
+// sligtly larger as jobs are poped from the todo stack in order and a
+// job close to the batchSize could force a very small batch to run.
+// Recommend minimum size: 256 + 64 = 320
+const batchSize = 320
+
+//
+// fetcher public interface
+//
+
+// startFetcher starts a new fetcher in the background
+func startFetcher(ctx context.Context, dserv ipld.DAGService) *fetcher {
+	log.Infof("fetcher: starting...")
+	f := &fetcher{
+		ctx:    ctx,
+		dserv:  dserv,
+		reqRes: make(chan *Shard),
+		result: make(chan result),
+		idle:   true,
+		done:   make(chan batchJob),
+		jobs:   make(map[*Shard]*job),
+	}
+	go f.mainLoop()
+	return f
+}
+
+// result contains the result of a job, see getResult
+type result struct {
+	vals map[string]*Shard
+	errs []error
+}
+
+// get gets the missing child shards for the hamt object.
+// The missing children for the passed in shard is returned.  The
+// children are then also retrieved in the background.  The result is
+// the result of the batch request and not just the single job.  In
+// particular, if the 'errs' field is empty the 'vals' of the result
+// is guaranteed to contain the all the missing child shards, but the
+// map may also contain child shards of other jobs in the batch
+func (f *fetcher) get(hamt *Shard) result {
+	f.reqRes <- hamt
+	res := <-f.result
+	return res
+}
+
+//
+// fetcher internals
+//
+
+type job struct {
+	id  *Shard
+	idx int /* index in the todo stack, an index of -1 means the job
+	   is already done or being worked on now */
+	cids []*cid.Cid
+	res  result
+}
+
+type jobStack struct {
+	c []*job
+}
+
+func (f *fetcher) mainLoop() {
+	var want *Shard
+	f.start = time.Now()
+	for {
+		select {
+		case id := <-f.reqRes:
+			if want != nil {
+				// programmer error
+				panic("fetcher: can not request more than one result at a time")
+			}
+			j, ok := f.jobs[id]
+			if !ok {
+				j = f.mainLoopAddJob(id)
+				if j == nil {
+					// no children that need to be retrieved
+					f.result <- result{vals: make(map[string]*Shard)}
+				}
+				if f.idle {
+					f.launch()
+				}
+			}
+			if j.res.vals != nil {
+				f.hits++
+				f.mainLoopSendResult(j)
+			} else {
+				if j.idx != -1 {
+					f.misses++
+					// move job to todoFirst so that it will be done on the
+					// next batch job
+					f.todo.remove(j)
+					f.todoFirst = j
+				} else {
+					f.nearMisses++
+				}
+				want = id
+			}
+		case bj := <-f.done:
+			f.doneCnt += len(bj.jobs)
+			f.cidCnt += len(bj.cids)
+			f.launch()
```

Comment:
I'm having trouble following the `launch()`/`idle` dynamic but this unconditional `launch()` call seems to be guaranteeing that we don't reach a deadlock state where the  `idle` is `false` and no one can call `launch()` to unblock it, right?